tuf/tuf.go

```diff
@@ -143,13 +143,16 @@ func (tr *Repo) RemoveBaseKeys(role string, keyIDs ...string) error {
 		}
 	}
 
-	// remove keys no longer in use by any roles
-	for k := range toDelete {
-		delete(tr.Root.Signed.Keys, k)
-		// remove the signing key from the cryptoservice if it
-		// isn't a root key. Root keys must be kept for rotation
-		// signing
-		if role != data.CanonicalRootRole {
+	// Remove keys no longer in use by any roles, except for root keys.
+	// Root private keys must be kept in tr.cryptoService to be able to sign
+	// for rotation, and root certificates must be kept in tr.Root.SignedKeys
+	// because we are not necessarily storing them elsewhere (tuf.Repo does not
+	// depend on certs.Manager, that is an upper layer), and without storing
+	// the certificates in their x509 form we are not able to do the
+	// util.CanonicalKeyID conversion.
+	if role != data.CanonicalRootRole {
+		for k := range toDelete {
+			delete(tr.Root.Signed.Keys, k)
```

Comment:
The *client* verifying signatures is holding a previous `root.json` file which contains the old root X509 public keys in `Signed.Role["root"]` (and which we want the client to migrate away from).

The *author* signing a new version the `root.json` may be completely done with that old X509 public key (for all we know, lacking more policy, it may now be several years after the author has migrated to a new trusted key from the old key we are talking about, or several generations of root keys after the old key, or the old root key may even known to be compromised)—but we still need to sign with the old key so that clients can rotate away from it. And in addition to just signing, we also need to store the X509 public key *somewhere* (a certificate can’t be regenerated from the raw non-X509 private key pair if we just know the ID=hash of the certificate).

For authors, we haven’t had designated a place to store the X509 public keys so far:
* It is not the `certs.Manager`, because the layering does not work well (as described in the comment), and because `certs.Manager` is used to manage the trusted certificates, and this old certificate may be known compromised and therefore not trusted = not suitable for `certs.Manager`.
* It is not a `trustmanager.KeyStore` because historically we haven’t been storing the generated X509 root public keys in there, so starting to use this location would require a migration step (though it could otherwise be a plausible candidate).

`root.json`’s `Signed.Keys` is actually the only place where the X509 public keys are currently being stored, so it seems easiest to designate this as the right place, and the implementation amounts to mostly the above hunk not deleting old keys which are already stored in there.

Also, in addition to just storing the X509 public keys, we need to be able to collect the set of X509 public keys which should be used to sign the `root.json`. Note that this is *not* `Signed.Roles["root"]` which contains keys trusted *for the future*, and should exclude the old known-compromised key. We could invent a special place to store the list of `root.json`-signing key IDs, but actually simply using the key IDs from the previous signature of `root.json` + the future trusted keys from `Signed.Roles["root"]` works just fine. Then the author’s copy of `root.json` becomes the canonical storage place for the root signing key IDs, so it seems natural to store the X509 public keys in the same place as well.